Source/WebCore/page/EventHandler.cpp

```diff
@@ -4416,17 +4416,31 @@ bool EventHandler::handleTextInputEvent(const String& text, Event* underlyingEve
 
     Ref frame = m_frame.get();
 
+    auto isEnterKeypress = [underlyingEvent]()
+    {
+        auto* event = dynamicDowncast<KeyboardEvent>(*underlyingEvent);
+        return event && event->type() == eventNames().keypressEvent && event->charCode() == '\r';
+    };
+
     EventTarget* target;
-    if (underlyingEvent)
+    if (underlyingEvent) {
         target = underlyingEvent->target();
-    else
+        // For an Enter keypress in an <input> element, return early — because, in that case,
+        // per the UI Events spec and webkit.org/b/54152, the textInput event must not be fired.
+        if (is<HTMLInputElement>(target) && isEnterKeypress())
+            return false;
+    } else
         target = eventTargetElementForDocument(frame->protectedDocument().get());
     if (!target)
         return false;
 
+    // The UI Events spec requires that textInput not be fired until after beforeinput is fired. So we create
+    // a textInput event here, but stop its propagation before dispatching it. And then we dispatch textInput
```

Comment:
Why are we firing textInput without propagation here? What's the point of this code?

Reply:
> ```
>    event->stopPropagation();
>    target->dispatchEvent(event);
> ```
> Why are we firing textInput without propagation here?

Fixed now — by instead directly calling `Editor::insertTextWithoutSendingTextEvent` with a nullptr for the `triggeringEvent` arg (rather than trying to preserve the existing `target->dispatchEvent(event)` call).

> What's the point of this code?

Per the _“If the event handler changed the selection, we may want to use a different selection that is contained in the event target”_ comment in the `Editor::insertTextWithoutSendingTextEvent`code (which we end up in after `EventDispatcher::dispatchEvent`), I’d been thinking the logic there for getting the selection still needed to check the target — and so, we needed to pass down an actual triggering event that it could get the target from. But if the event handler here isn’t changing the selection, then that code doesn’t need an actual event to get the target from.

Reply:
> Fixed now — by instead directly calling `Editor::insertTextWithoutSendingTextEvent` with a nullptr for the `triggeringEvent` arg (rather than trying to preserve the existing `target->dispatchEvent(event)` call).

Seems I spoke too soon. That replacement  causes [regressions with several test cases](https://ews-build.webkit.org/#/builders/36/builds/41661) – including the `fast/forms/focus-change-on-keypress.html` test — and I recall now seeing those same regressions a few weeks earlier when I was first trying out solutions for avoiding the event dispatch.

> > What's the point of this code?
> 
> …in the `Editor::insertTextWithoutSendingTextEvent`code… I’d been thinking the logic there for getting the selection still needed to check the target… but if the event handler here isn’t changing the selection, then that code doesn’t need to check the target

Seems I may have been wrong about that not being needed. I haven’t yet looked in detail at the exact causes of all the regressions, but so far I did step through the `fast/forms/focus-change-on-keypress.html` execution — and the cause in that case does in fact seem to be that we’re skipping the check of the target to get the selection.

So for now, I have reverted to what I had earlier: preserving the `target->dispatchEvent(event)` call — so that the selection-setting code in `Editor::insertTextWithoutSendingTextEvent` can continue working as expected — but then stopping propagation of the event.

I realize that’s kludgy but it’s not clear to me how to otherwise avoid the regressions and get the behavior conforming without reworking a bunch more of the existing code.

But maybe I’m overlooking something and there’s some simpler change we could make to avoid the kludge.